search-provider/yelp-search-provider.c

```diff
@@ -0,0 +1,568 @@
+                        gchar                    **terms,
+                        guint32                    timestamp,
+                        gpointer                   user_data)
+{
+    return yelp_launch (result, user_data, invocation, "show-page");
+}
+
+static void
+search_provider_app_dispose (GObject *obj)
+{
+    YelpSearchProviderApp *self = YELP_SEARCH_PROVIDER_APP (obj);
+
+    g_clear_object (&self->skeleton);
+    g_clear_pointer (&self->page_data_hash_map, g_hash_table_unref);
+
+    G_OBJECT_CLASS (yelp_search_provider_app_parent_class)->dispose (obj);
+}
+
+static void
+preload_data_cb (YelpDocument          *document,
+                 YelpDocumentSignal     signal,
+                 YelpSearchProviderApp *app,
+                 GError                *error)
+{
+    YelpSearchProviderApp *self = YELP_SEARCH_PROVIDER_APP (app);
+    gchar **page_ids = yelp_document_list_page_ids (document);
+    gchar **iter;
+
+    if (signal == YELP_DOCUMENT_SIGNAL_ERROR) {
+        if (error)
+            g_warning ("error: %s", error->message);
+        return;
+    }
+
+    if (signal == YELP_DOCUMENT_SIGNAL_CONTENTS)
+        return;
+
+    if (signal == YELP_DOCUMENT_SIGNAL_INFO) {
+        for (iter = page_ids; *iter; iter++) {
+            gchar *page_id = *iter;
+            PageData *data = g_new0 (PageData, 1);
+            gchar *icon_string = yelp_document_get_page_icon (document, page_id);
+            GIcon *icon = g_icon_new_for_string (icon_string, NULL);
+            g_free (icon_string);
+
+            data = page_data_new_steal (yelp_document_get_page_title (document, page_id), 
+                                        yelp_document_get_page_desc (document, page_id), 
+                                        icon);
+            g_hash_table_insert (app->page_data_hash_map, page_id, data);
+        }
+
+        g_application_release (G_APPLICATION (self));
+        g_free (page_ids);
+    }
+}
+
+static void
+uri_resolved_cb (YelpUri               *uri,
+                 YelpSearchProviderApp *app)
+{
+    YelpDocument *doc;
+    gchar *page_id;
+
+    doc = yelp_document_get_for_uri (uri);
+    g_signal_handler_disconnect (uri, app->uri_resolve_id);
+    app->uri_resolve_id = 0;
+    page_id = yelp_uri_get_page_id (uri);
+    yelp_document_request_page (doc,
+                                page_id,
+                                NULL,
+                                (YelpDocumentCallback) preload_data_cb,
+                                app,
+                                NULL);
+
+    g_object_unref (doc);
+}
+
+static void
+search_provider_app_startup (GApplication *app)
+{
+    YelpSearchProviderApp *self = YELP_SEARCH_PROVIDER_APP (app);
+    YelpUri *base_uri;
+
+    G_APPLICATION_CLASS (yelp_search_provider_app_parent_class)->startup (app);
+
+    if (g_getenv ("YELP_SEARCH_PROVIDER_PERSIST") != NULL)
+        g_application_hold (app);
+
+    base_uri = yelp_uri_new ("help:gnome-help");
```

Comment:
What about my earlier suggestion about exporting the `DEFAULT_URI` macro in yelp-uri.h as `YELP_GNOME_HELP_URI`, so you could use it here?